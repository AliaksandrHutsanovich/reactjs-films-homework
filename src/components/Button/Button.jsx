import React, { memo, useMemo } from 'react';
import PropTypes from 'prop-types';
import { Button as AntdButton } from 'antd';

import '!style-loader!css-loader!sass-loader!antd/dist/antd.css'; //eslint-disable-line

import styles from './Button.scss';

const ANTD_BUTTON_TYPE_CONFIG = {
  outlined: 'default',
  filled: 'primary',
};

const Button = ({
  disabled,
  size,
  type,
  onClick,
  title,
}) => {
  const className = useMemo(() => (type === 'outlined' ? styles.outlined : ''), [type]);
  const antdType = ANTD_BUTTON_TYPE_CONFIG[type];
  return (
    <AntdButton
      onClick={onClick}
      disabled={disabled}
      size={size}
      className={className}
      type={antdType}
    >
      {title}
    </AntdButton>
  );
};

Button.propTypes = {
  disabled: PropTypes.bool,
  type: PropTypes.oneOf([
    'filled',
    'outlined',
  ]),
  onClick: PropTypes.func.isRequired,
  size: PropTypes.oneOf([
    'large',
    'middle',
    'small',
  ]),
  title: PropTypes.string.isRequired,
};

Button.defaultProps = {
  disabled: false,
  type: 'filled',
  size: 'middle',
};

export default memo(Button);
