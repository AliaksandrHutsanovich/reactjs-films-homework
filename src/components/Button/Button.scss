.outlined {
  border: 1px solid white !important;
  color: white !important;
  background-color: transparent !important;
}
