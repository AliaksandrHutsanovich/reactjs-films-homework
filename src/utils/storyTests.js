const storyTests = (stories, renderer) => {
  Object.keys(stories).forEach((storyKey) => {
    if (typeof stories[storyKey] !== 'function') {
      return;
    }
    it(`${storyKey} should be rendered properly`, () => {
      renderer.render(stories[storyKey]());
      const result = renderer.getRenderOutput();

      expect(result).toMatchSnapshot();
    });
  });
};

export default storyTests;
