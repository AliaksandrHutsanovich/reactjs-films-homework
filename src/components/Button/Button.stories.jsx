import React from 'react';
import { action } from '@storybook/addon-actions';
import Button from './Button';
import image from '../../assets/images/story-background.png';

export default {
  component: Button,
  title: 'components/Button',
};

export const outlined = () => (
  <Button
    title="Click me"
    onClick={action('the default button is clicked')}
    type="outlined"
  />
);
outlined.story = {
  decorators: [
    (story) => (
      <div
        style={{
          background: `url(${image}) no-repeat`,
          padding: '20px',
          width: '200px',
          height: '200px',
        }}
      >
        {story()}
      </div>
    ),
  ],
};

export const filled = () => (
  <Button
    title="Click me"
    onClick={action('the primary button is clicked')}
  />
);
