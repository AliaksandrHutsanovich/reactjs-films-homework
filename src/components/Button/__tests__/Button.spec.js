import ShallowRenderer from 'react-test-renderer/shallow';
import storyTests from '../../../utils/storyTests';
import * as stories from '../Button.stories';

describe('<Button />', () => {
  const renderer = new ShallowRenderer();

  storyTests(stories, renderer);
});
