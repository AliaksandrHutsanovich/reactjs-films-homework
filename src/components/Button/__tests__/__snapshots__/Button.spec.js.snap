// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> filled should be rendered properly 1`] = `
<ForwardRef(InternalButton)
  block={false}
  className=""
  ghost={false}
  htmlType="button"
  loading={false}
  onClick={[Function]}
>
  Click me
</ForwardRef(InternalButton)>
`;

exports[`<Button /> outlined should be rendered properly 1`] = `
<ForwardRef(InternalButton)
  block={false}
  className="outlined"
  ghost={false}
  htmlType="button"
  loading={false}
  onClick={[Function]}
  type="default"
>
  Click me
</ForwardRef(InternalButton)>
`;
